Sources/PostgresNIO/New/PSQLRows.swift

```diff
@@ -53,48 +70,93 @@ final class PSQLRows {
             }
         }
         
-        assert(self.downstreamState == .consuming)
+        guard case .consuming = self.downstreamState else {
+            preconditionFailure("Invalid state")
+        }
         
         switch self.upstreamState {
-        case .streaming(let upstreamNext, _):
-            return upstreamNext().map { payload -> Row? in
-                self.downstreamState = .consuming
-                switch payload {
-                case .row(let data):
-                    return Row(data: data, lookupTable: self.lookupTable, columns: self.rowDescription, jsonDecoder: self.jsonDecoder)
-                case .complete(var buffer, let commandTag):
-                    if let data = buffer.popFirst() {
-                        self.upstreamState = .finished(remaining: buffer, commandTag: commandTag)
-                        return Row(data: data, lookupTable: self.lookupTable, columns: self.rowDescription, jsonDecoder: self.jsonDecoder)
-                    }
-                    
-                    self.upstreamState = .consumed(.success(commandTag))
-                    return nil
-                }
-            }.flatMapErrorThrowing { error in
-                // if we have an error upstream that, we pass through here, we need to set
-                // our internal state
-                self.upstreamState = .consumed(.failure(error))
-                throw error
-            }
+        case .streaming(var buffer, let dataSource):
+            self.upstreamState = .modifying
             
-        case .finished(remaining: var buffer, commandTag: let commandTag):
-            self.downstreamState = .consuming
             if let data = buffer.popFirst() {
-                self.upstreamState = .finished(remaining: buffer, commandTag: commandTag)
+                self.downstreamState = .consuming
+                self.upstreamState = .streaming(buffer: buffer, dataSource: dataSource)
                 let row = Row(data: data, lookupTable: self.lookupTable, columns: self.rowDescription, jsonDecoder: self.jsonDecoder)
                 return self.eventLoop.makeSucceededFuture(row)
             }
             
+            let promise = self.eventLoop.makePromise(of: Row?.self)
+            self.downstreamState = .waitingForNext(promise)
+            self.upstreamState = .streaming(buffer: buffer, dataSource: dataSource)
+            dataSource.request()
```

Comment:
This is the invocation, where the PSQLRowStream asks for more rows. We could change this to:

```swift
context.channel.triggerUserOutboundEvent(PSQLRowStreamEvent.requestMoreRows, promise: nil)
```